ip_proofs/src/tipa/structured_scalar_message.rs

```diff
@@ -0,0 +1,205 @@
+use algebra::{curves::PairingEngine, fields::Field};
+use digest::Digest;
+use num_traits::identities::One;
+use std::ops::MulAssign;
+
+use crate::{
+    gipa::GIPA,
+    tipa::{TIPACompatibleSetup, TIPAProof, VerifierSRS, SRS, TIPA},
+    Error,
+};
+use dh_commitments::DoublyHomomorphicCommitment;
+use inner_products::InnerProduct;
+
+//TODO: Properly generalize the non-committed message approach of SIPP and MIPP to GIPA
+//TODO: Structured message is a special case of the non-committed message and does not rely on TIPA
+//TODO: Can support structured group element messages as well as structured scalar messages
+
+// TODO: Currently implemented by completely reusing TIPA, which means wasted prover and verifier effort on B
```

Comment:
Maybe you can extract out the common code into a method on TIPA, and then use that method in both TIPA and here.

Reply:
Yeah, I think something like that could work. The problem with using TIPA black-box as I am right now is that it attempts to calculate the structured KZG polynomial commitment proof for both `left` and `right`. We would want to extract the polynomial commitment proving and verification code from the main prove and verify methods, so that they can be called separately for `left` and `right`.

Then to do TIPA for `left` and a structured scalar message for `right`, we can first run GIPA using a `PlaceholderCommitment` that has null commitment keys and null commitments. Then run the TIPA polynomial commitment code on just `left`.

I haven't made this change yet, but commenting here to make note of this as a possible plan forward.